documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,809 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+One final point to note is that throughout the rest of the proposal the following type will be used:
+
+```cs
+public struct InputReadOnlyList<T> : IReadOnlyList<T>
+{
+    public InputReadOnlyList(IReadOnlyList<T> other);
+}
+```
+
+The Silk.NET team wishes to reserve the right to add more constructors to this type as it sees fit.
+
+This exists so that, should the Silk.NET choose to, we can optimize the lookup of elements while ensuring things like indexers are inlined and don't result in a virtual call if our implementation allows us to do so.
+
+**INFORMATIVE TEXT:** For example, for joystick and mouse buttons we could use a fixed-sized bit buffer where each bit represents an individual button: 1 for pressed, 0 for unpressed. But for something like keyboard input where there are a large amount of keys, we can't do that and will likely use `Memory<T>` instead.
+
+# Mouse Input
+
+As discussed earlier, the interface will be very simple.
+
+```cs
+public interface IMouse : IInputDevice
+{
+    MouseState State { get; }
+    ICursorConfiguration Cursor { get; }
+    void SetPosition(Vector2 pos);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`Cursor` contains the cursor configuration. This isn't actually state that the end user can change, and has been made an interface rather than a state struct accordingly.
+
+`SetPosition` allows moving the mouse cursor without the end user physically moving their mouse. Please note that this does not immediately update `State` with the new value - the changes will be reflected next time `IInputBackend.Update` is called.
+
+The device state returned by `State` fills out the following structure:
+
+```cs
+public readonly record struct MouseState
+(
+    MouseButtonState Buttons,
+    Vector2 Position,
+    Vector2 WheelPosition
+);
+```
+
+`MouseButtonState` is defined as:
+```cs
+public readonly record struct MouseButtonState
+(
+    InputReadOnlyList<MouseButton> Down
+)
+{
+    public bool this[MouseButton btn] { get; }
+    public InputReadOnlyList<MouseButton> Up { get; }
+}
+```
+
+The indexer returns `true` if a particular button is pressed, false otherwise. If the developer wishes to enumerate the button state, they must explicitly enumerate through even the `Down` buttons or `Up` buttons.
+
+`Up` and the indexer will be implemented in terms of `Down`, which is the only property that a backend will need to set.
+
+Changes to `MouseState` also have matching actor methods which are subject to the actor method rules i.e. the backend should call them in the order in which the backend received the events where possible etc (read the Actors section).
+
+```cs
+public interface IMouseActor : IActor
+{
+    void HandleButtonDown(ButtonEvent<IMouse, MouseButton> @event);
+    void HandleButtonUp(ButtonEvent<IMouse, MouseButton> @event);
+    void HandleCursorMove(AxisEvent<IMouse, Vector2> @event);
+    void HandleScroll(AxisEvent<IMouse, Vector2> @event);
```

Comment:
Mouse button events should come with the modifier keys that are pressed at the time of the event, like in the proposal for Enhanced Input Events. Same goes for keyboard keyup/keydown.

Reply:
Got it. I suppose I should make all of the events non-generic so we can be free to add extra properties to them later.

Reply:
That does sound nice

Reply:
@Perksey I think we should add the key modifiers argument not just to ButtonDown/ButtonUp but also to Click/DoubleClick

Reply:
As discussed in Discord [here](https://discord.com/channels/521092042781229087/587346162802229298/941082429756690452), scrapping this.